format-common/src/main/java/io/cdap/plugin/format/charset/CharsetTransformingLineRecordReader.java

```diff
@@ -0,0 +1,222 @@
+/*
+ * Copyright © 2020 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.plugin.format.charset;
+
+import io.cdap.plugin.format.charset.fixedlength.FixedLengthCharset;
+import io.cdap.plugin.format.charset.fixedlength.FixedLengthCharsetTransformingCodec;
+import io.cdap.plugin.format.charset.fixedlength.FixedLengthCharsetTransformingDecompressorStream;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FSDataInputStream;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.fs.Seekable;
+import org.apache.hadoop.io.LongWritable;
+import org.apache.hadoop.io.Text;
+import org.apache.hadoop.io.compress.CodecPool;
+import org.apache.hadoop.io.compress.Decompressor;
+import org.apache.hadoop.io.compress.SplitCompressionInputStream;
+import org.apache.hadoop.io.compress.SplittableCompressionCodec;
+import org.apache.hadoop.mapreduce.InputSplit;
+import org.apache.hadoop.mapreduce.RecordReader;
+import org.apache.hadoop.mapreduce.TaskAttemptContext;
+import org.apache.hadoop.mapreduce.lib.input.CompressedSplitLineReader;
+import org.apache.hadoop.mapreduce.lib.input.FileSplit;
+import org.apache.hadoop.mapreduce.lib.input.LineRecordReader;
+import org.apache.hadoop.mapreduce.lib.input.SplitLineReader;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+
+/**
+ * Copy of Hadoop's LineRecordReader (Hadoop Version 2.3.0). The reason we copy this class is to modify some behaviors
+ * related to the creation of the decompressor. This also allows us to implement
```

Comment:
What is modified and why is needed?

Reply:
@chtyim The problem with variable with encoding is that the file partitions in Spark are based on file position in bytes, meaning we might end up in a scenario where the partition starts in the middle of a character.

If this is the case, we would be in a position where the input stream cannot be decoded correctly. Essentially, trying to read from this input stream as a String will return an invalid string.

For Fixed Length charsets, my implementations makes sure to align the partition boundaries with full characters: https://github.com/cdapio/hydrator-plugins/blob/83427f7002b016044f5f4606d998b92511c92ceb/format-common/src/main/java/io/cdap/plugin/format/charset/fixedlength/FixedLengthCharsetTransformingCodec.java#L110-L124

The only flexible solution I see for variable-length encodings is to run the decoding from the start of the file in order to make sure we are always aligned to a character boundary.

Reply:
@chtyim The original implementation of Hadoop's LineRecordReader uses the file name to determine which codec to implement:

https://github.com/facebookarchive/hadoop-20/blob/master/src/mapred/org/apache/hadoop/mapred/LineRecordReader.java#L92

https://github.com/facebookarchive/hadoop-20/blob/master/src/core/org/apache/hadoop/io/compress/CompressionCodecFactory.java#L160-L175

This is not particularly useful in our use case as the file extension could be anything.

In order to go around this limitation, I copied this class and simplified the code as we can assume the input is "compressed" and we need to "decompress" (apply the charset conversion) as we read the file.

Reply:
It is controlled by config 

https://github.com/apache/hadoop/blob/branch-2.9.2/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/io/compress/CompressionCodecFactory.java#L124

Reply:
Can we do it without copying this class?

Reply:
I'll work on extending the LineRecordReader class and override this functionality instead of copying the whole class.

Reply:
@chtyim I don't think the codec-by-configuration approach will work for us because the default implementation of the LineRecordReader class always invokes this function:

https://github.com/apache/hadoop/blob/branch-2.9.2/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/io/compress/CompressionCodecFactory.java#L197-L213

This means that the codec is matched based on file extension, which is not a desirable behavior in my case.

Reply:
@tivv You might be right. Let me give it a try. I had not considered that `headMap` using an empty string as key will just match every record in the map.

Reply:
@tivv @chtyim After going through this attempt, I discovered a potential issue with using the default configuration:

If you look at the `setCodecClasses` method, the CompressionCodecFactory will also load CompressionCodecs in the classpath using the ServiceLoader: https://hadoop.apache.org/docs/r2.7.3/api/org/apache/hadoop/io/compress/CompressionCodecFactory.html#setCodecClasses(org.apache.hadoop.conf.Configuration,%20java.util.List).

This means the `headMap` method call will always return multiple potential candidates. The default implementation then, picks the last result in this new map:

![image](https://user-images.githubusercontent.com/7290959/102375906-0e709880-3f91-11eb-849b-ad4e999e851d.png)

I don't think we can get around this limitation (unless we override every single possible file extension with our codec) but this does not seem like a good idea.

Reply:
Yep, it would be very hacky to work around ServiceLoader. We would need to change the context classloader. Not worth it in my opinion.